src/driver-asus.c

```diff
@@ -0,0 +1,517 @@
+/*
+ * Copyright (C) 2021 Kyoken, kyoken@kyoken.ninja
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include "config.h"
+#include <assert.h>
+#include <errno.h>
+#include <libevdev/libevdev.h>
+#include <linux/input.h>
+#include <math.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "libratbag-private.h"
+#include "libratbag-hidraw.h"
+#include "libratbag-data.h"
+
+#include "asus.h"
+
+/* ButtonMapping configuration property defaults, stores indices for ASUS_BUTTON_MAPPING */
+static int8_t ASUS_BUTTON_MAPPING_INDEX[] = {
+	0,  /* left */
+	1,  /* right (button 3 in xev) */
+	2,  /* middle (button 2 in xev) */
+	6,  /* wheel up */
+	7,  /* wheel down */
+	5,  /* DPI */
+	3,  /* backward */
+	4,  /* forward */
+	-1,  /* placeholder */
+	-1,  /* placeholder */
+	-1,  /* placeholder */
+	-1,  /* placeholder */
+	-1,  /* placeholder */
+	-1,  /* placeholder */
+	-1,  /* placeholder */
+	-1,  /* placeholder */
+};
+
+static unsigned int ASUS_LED_MODE[] = {
+	RATBAG_LED_ON,
+	RATBAG_LED_BREATHING,
+	RATBAG_LED_CYCLE,
+	RATBAG_LED_ON,  /* rainbow wave */
+	RATBAG_LED_ON,  /* reactive - react to clicks */
+	RATBAG_LED_ON,  /* custom - depends on mouse type */
+	RATBAG_LED_ON,  /* battery - battery indicator */
+};
+
+struct asus_data {
+	uint8_t is_ready;
+} __attribute__((packed));
```

Comment:
I don't think you need `__attribute__((packed))` as it's not some data we receive from the mouse, it's something we define and use ourselves.